server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala

```diff
@@ -23,45 +23,65 @@ import java.net.{URI, URL}
+      websocketPrefixString = Option(configuration.getString(WEBSOCKET_URL_PREFIX_PROPERTY)).getOrElse(WEBSOCKET_URL_PREFIX_DEFAULT),
+      dynamicJmapPrefixResolutionEnabled = configuration.getBoolean(DYNAMIC_JMAP_PREFIX_RESOLUTION_ENABLED_PROPERTY, false),
+      maxUploadSize = Option(configuration.getString(UPLOAD_LIMIT_PROPERTY, null))
         .map(Size.parse)
         .map(MaxSizeUpload.of(_).get)
-        .getOrElse(UPLOAD_LIMIT_30_MB),
-      maxTimeoutSeconds = Optional.ofNullable(configuration.getInteger("webpush.maxTimeoutSeconds", null)).map(Integer2int).toScala,
-      maxConnections = Optional.ofNullable(configuration.getInteger("webpush.maxConnections", null)).map(Integer2int).toScala)
-  }
+        .getOrElse(UPLOAD_LIMIT_DEFAULT),
+      maxTimeoutSeconds = Optional.ofNullable(configuration.getInteger(WEB_PUSH_MAX_TIMEOUT_SECONDS_PROPERTY, null)).map(Integer2int).toScala,
+      maxConnections = Optional.ofNullable(configuration.getInteger(WEB_PUSH_MAX_CONNECTIONS_PROPERTY, null)).map(Integer2int).toScala)
 }
 
-case class JmapRfc8621Configuration(urlPrefixString: String, websocketPrefixString: String,
-                                    maxUploadSize: MaxSizeUpload = UPLOAD_LIMIT_30_MB,
+case class JmapRfc8621Configuration(urlPrefixString: String,
+                                    websocketPrefixString: String,
+                                    dynamicJmapPrefixResolutionEnabled: Boolean = false,
+                                    maxUploadSize: MaxSizeUpload = UPLOAD_LIMIT_DEFAULT,
                                     maxTimeoutSeconds: Option[Int] = None,
                                     maxConnections: Option[Int] = None) {
-  val urlPrefix: URL = new URL(urlPrefixString)
-  val apiUrl: URL = new URL(s"$urlPrefixString/jmap")
-  val downloadUrl: URL = new URL(urlPrefixString + "/download/{accountId}/{blobId}?type={type}&name={name}")
-  val uploadUrl: URL = new URL(s"$urlPrefixString/upload/{accountId}")
-  val eventSourceUrl: URL = new URL(s"$urlPrefixString/eventSource?types={types}&closeAfter={closeafter}&ping={ping}")
-  val webSocketUrl: URI = new URI(s"$websocketPrefixString/jmap/ws")
 
+  def getURLPrefix(urlPrefixRequest: Option[URL]): String = Some(dynamicJmapPrefixResolutionEnabled)
+    .filter(enabled => enabled)
+    .flatMap(_ => urlPrefixRequest.map(_.toString))
+    .getOrElse(urlPrefixString)
+
+  def apiUrl(urlPrefixRequest: Option[URL] ): URL = new URL(s"${getURLPrefix(urlPrefixRequest)}/jmap")
+
+  def downloadUrl(urlPrefixRequest: Option[URL]): URL = new URL(s"${getURLPrefix(urlPrefixRequest)}/download/{accountId}/{blobId}?type={type}&name={name}")
+
+  def uploadUrl(urlPrefixRequest: Option[URL]): URL = new URL(s"${getURLPrefix(urlPrefixRequest)}/upload/{accountId}")
+
+  val webSocketUrl: URI = new URI(s"$websocketPrefixString/jmap/ws")
+  val eventSourceUrl: URL = new URL(s"$urlPrefixString/eventSource?types={types}&closeAfter={closeafter}&ping={ping}")
```

Comment:
Is a configuration POJO the right place for advanced resolution? Shouldn't an other entity do this logic relying only on the prefix shiped by the configuration?

Reply:
I created the `JmapUrlEndpointResolver` class for this